agent/consul/client_test.go

```diff
@@ -665,3 +667,25 @@ func TestClient_Encrypted(t *testing.T) {
 		t.Fatalf("should be encrypted")
 	}
 }
+
+func TestClient_Reload(t *testing.T) {
+	t.Parallel()
+	dir1, c := testClientWithConfig(t, func(c *Config) {
+		c.RPCRate = 500
+		c.RPCMaxBurst = 5000
+	})
+	defer os.RemoveAll(dir1)
+	defer c.Shutdown()
+
+	limiter := c.rpcLimiter.Load().(*rate.Limiter)
+	require.Equal(t, rate.Limit(500), limiter.Limit())
+	require.Equal(t, 5000, limiter.Burst())
+
+	c.config.RPCRate = 1000
+	c.config.RPCMaxBurst = 10000
+
+	require.NoError(t, c.ReloadConfig(c.config))
```

Comment:
This all looks good to me but I wonder if we should have 3 different goroutines all spawn and run this in a loop and verify that `go test -race` doesn't pick up any concurrency issues?

Reply:
ReloadConfig can't be called concurrently. At least as the code stands today.

Client.ReloadConfig is only called from Agent.ReloadConfig which in turn is only called from cmd.handleReload (command/agent/agent.go) and finally handleReload is only called from a loop in cmd.run which is processing the chan of signals. As it can only handle 1 signal at a time, ReloadConfig can therefore not be called concurrently so there can be no race. I have no doubt that if it was called concurrently there would be several problems (like watches, services etc could be loaded multiple times)

Reply:
Oh OK. Maybe I confused the calls, I was mostly wanting to validate the `atomic.Value` stuff - I guess that is about calling this concurrently with RPC client reading the values so perhaps the valid test would be to start a goroutine (or 10) making RPCs while calling this in a loop or something.

It's not a blocker, it just might be an extra sanity check that there aren't any obvious concurrency issues.

Reply:
What would that really test? To me it looks like that would be testing that atomc.Value.(Load/Store) do what they say they do. And since that is an upstream library (and builtin to the language) I would say that testing that is outside the scope of Consul testing and would only serve to lengthen test times. We should be able to assume that those atomic primitives work as expected. Because the whole limiter is now stored within the value it is impossible to access it without a load/store and since we don't do any write operations into an existing limiter during reload (but rather create a new limiter) there can be no race. Additionally the limiter looks to be thread safe in that if you did set new limits or perform other operations it will lock itself internally (which is how we can use one limiter with multiple go routines). Looks to be = I read through the code. I don't think they expressly say in the package documentation that this is the case.

Reply:
> We should be able to assume that those atomic primitives work as expected. 

It's often the case that atomic values are misused. My suggestion was to guard against any of the common pitfalls of using them!

I don't think it's critical so merge away but in general go's race detector is easy to use and can often catch subtle and common mistakes/bad assumptions in how `sync` primitives are used. 😄 .